Altair P2P (#2244)
ethereum/eth2.0-specs: specs/altair/p2p-interface.md

# Ethereum Altair networking specification

This document contains the networking specification for Ethereum 2.0 clients added during the Altair deployment.
This document should be viewed as additive to the [document from Phase 0](../phase0/p2p-interface.md) and will be referred to as the "phase 0 document" hereafter.
Readers should understand the phase 0 document and use it as a basis to understand the changes outlined in this document.

> Altair adds new messages, topics and data to the Req-Resp, Gossip and Discovery domain. Some phase0 features will be deprecated, but not removed immediately.

Reviewer: ```suggestion
Altair adds new messages, topics and data to the Req-Resp, Gossip and Discovery domain. Some phase 0 features will be deprecated, but not removed immediately.
```

but could go either way as long as we are consistent

Definitions of these new types can be found in the [Altair validator guide](./validator.md#containers).

The `ForkDigestValue` path segment of the topic separates the old and the new `beacon_block` topics.

#### Global topics

Altair changes the type of the global beacon block topic and adds one global topic to propagate partially aggregated sync committee signatures to all potential proposers of beacon blocks.

##### `beacon_block`

> The existing specification for this topic do not change from the Phase 0 document, 

Reviewer: ```suggestion
The existing specifications for this topic do not change from the Phase 0 document, 
```

Aggregated sync committee signatures are packaged into (signed) `SyncCommitteeContribution` along with proofs and gossiped to the `sync_committee_contribution_and_proof` topic.

### Transitioning the gossip

With any fork, the fork version, and thus the `ForkDigestValue`, change.

> Message types are unique per topic, and so for a smooth transition a node will have to subscribe to both the old and new topics.

Reviewer 2: ```suggestion
Message types are unique per topic, and so for a smooth transition a node must temporarily subscribe to both the old and new topics.
```

> All remaining topics are updated. In advance of the fork, a node SHOULD subscribe to the post-fork variants of the topics.

Reviewer: ```suggestion
All remaining topics are updated with the new `ForkDigestValue`. In advance of the fork, a node SHOULD subscribe to the post-fork variants of the topics.
```

this is a bit more clear imo

Author: I think the "remaining" part is confusing. It's more like "those topics names that stay in use after the fork update". I'll try to fix it

> All remaining topics are updated. In advance of the fork, a node SHOULD subscribe to the post-fork variants of the topics.

Author: ```suggestion
The topics that are not removed in a fork are updated with a new `ForkDigestValue`. In advance of the fork, a node SHOULD subscribe to the post-fork variants of the topics.
```

Subscriptions are expected to be well-received, all updated nodes should subscribe as well.
Topic-meshes can be grafted quickly as the nodes are already connected and exchanging gossip control messages.

Messages SHOULD NOT be re-broadcast from one fork to the other.

> 

Reviewer: ```suggestion
A node's behavior before the fork and after the fork are as follows:
```

I think adding a bit of prose here to guide the reader will help. I was a bit confused when we just dropped into pre and post and just an additional sentence that this is the expected behavior in each era is helpful

Pre-fork:
- Peers who propagate messages on the post-fork topics MAY be scored negatively proportionally to time till fork,
  to account for clock discrepancy.
- Messages can be IGNORED on the post-fork topics, with a `MAXIMUM_GOSSIP_CLOCK_DISPARITY` margin.

Post-fork:
- Peers who propagate messages on the pre-fork topics MUST NOT be scored negatively. Lagging IWANT may force them to.

> - Messages on pre and post-fork variants of topics share application-level caches.

Reviewer: good call

  E.g. an attestation on the both the old and new topic is ignored like any duplicate.

> - Two epochs after the fork, pre-fork topics SHOULD be unsubscribed from. This is well after the configured `seen_ttl`.

Reviewer: just noting that we are tearing down the topics so no worries if suddenly a large portion of the network drops a topic all at once (vs. subnet stability elsewhere where it would be an issue); `seen_ttl` is 1 epoch so 2 epochs should be plenty of buffer, even accounting for clock disparity

## The Req/Resp domain

### Req-Resp interaction

> An additional `<context-bytes>` field is introduced to the `response_chunk` as defined in the phase0 document:

Reviewer: ```suggestion
An additional `<context-bytes>` field is introduced to the `response_chunk` as defined in the phase 0 document:
```

either way, lets just be consistent

```
response_chunk  ::= <result> | <context-bytes> | <encoding-dependent-header> | <encoded-payload>
```

> All phase0 methods are compatible: `<context-bytes>` is empty by default.

Reviewer: ```suggestion
All phase 0 methods are compatible: `<context-bytes>` is empty by default.
```

ditto

On a non-zero `<result>` with `ErrorMessage` payload, the `<context-bytes>` is also empty.

In Altair and later forks, `<context-bytes>` functions as a short meta-data,
defined per req-resp method, and can parametrize the payload decoder.

#### `ForkDigest`-context

Starting with Altair, and in future forks, SSZ type definitions may change.
For this common case, we define the `ForkDigest`-context:

A fixed-width 4 byte `<context-bytes>`, set to the `ForkDigest` matching the chunk:
 `compute_fork_digest(fork_version, genesis_validators_root)`.

### Messages

> #### BeaconBlocksByRange v2

Reviewer 3: How does adding the `context` metadata work here ? They are only included in response packets rather than requests. Is the goal for consumers to be aware of the particular fork version of the response payload ?

Author: The version bump is to distinguish between an empty `context`, and a `ForkDigest` used in v2. The format is "compatible" in the sense that we can define the format of the old v1 methods as a subset of the new format, by simply assuming an empty context. But the actual change of a method is breaking, hence a v2.

The idea is that everyone can update their Req-Resp stack with this small change, and keep supporting the old formats almost effortlessly. Most methods don't need a v2 with Altair. Just in the case of these two beacon block methods, the fork digest can now be read before touching the SSZ decoder, which should make it easy to handle different block types.

Regarding implementation, the verbose way of describing it would be:
1. Add a `CodecContext` interface/base-type/whatever preferred.
2. Parametrize the response writer with `CodecContext`
3. Specify `EmptyContext` as a subtype/implementation/whatever of `CodecContext`, with a read/write method, that is a no-op.
4. Specify that all v1 request handlers have an `EmptyContext` as context type
5. Specify that all v1 response handlers have an `EmptyContext` as context type
6. Specify `ForkDigestContext` as subtype/implementation/whatever of `CodecContext`, with a read/write method, that reads/writes 4 bytes into/from a fork digest.
6. Define the new v2 block request handlers in terms of the old v1 handlers, but parametrized with `ForkDigestContext`. The chunk writer can switch on the type of context, and when it's `ForkDigestContext`, it writes the fork digest corresponding to the block.
7. Define the new v2 block response handlers in terms of the old v1 handlers, but parametrized with `ForkDigestContext`. The chunk reader first reads the expected 4 byte fork digest, validates if it's a known fork, and then continues to decode the chunk into the expected type of block.

We can also rename the req-resp `CodecContext` to `CodecParameter` or something if it's confusing in Go.
